### PR TITLE
fix: correctly parse documentation comments

### LINKED ISSUE #258: [Parser] Comments and WhiteSpace
```
///
/// Foo
/// fsefa
rel R(x: Int)

/// This is A
///
rel A(x: Int)

//fs asd fas

/* fsd fasd */

///
/// This is B
///
rel B(x: Int)

///
/// This is C
///
rel C(x: Int)
```

### PR DESCRIPTION
**Pull request by JonathanStarup**
Fixes #258.

### CODE REVIEW

**Comment by JonathanStarup** [comment]
<details>
 <summary>Throughput Benchmarks</summary>
**This PR:**
```
Throughput (best): 10.717 lines/sec (with 8 threads.)

  min:  3.063, max: 10.717, avg:  8.706, median: 10.024

  The highest throughput was in iteration: 8 (out of 14).
  The ratio between the best and worst iteration was: 3,5x.

Finished 14 iterations on 58.966 lines of code in 106 seconds.
```
**Master:**
```
====================== Flix Compiler Throughput ======================

Throughput (best): 10.721 lines/sec (with 8 threads.)

  min:  3.047, max: 10.721, avg:  8.932, median:  9.665

  The highest throughput was in iteration: 10 (out of 14).
  The ratio between the best and worst iteration was: 3,5x.

Finished 14 iterations on 58.966 lines of code in 103 seconds.
```
</details>

**Comment by JonathanStarup** [comment]
doc comments are seperated by anything other than space and tab
```
/// This is 

/// not a single doc comment

/// this is
     /// a single
  /// doc comment
```
And a defs/classes/etc. use the last defined doc comment, so this is valid
```
/// doc comment

// something
/* somehitng */
def f(): Unit = ()
```

**Comment by magnus-madsen** [comment]
This looks more reasonable 👍

**Comment by magnus-madsen** [comment]
Could you try `--Xbenchmark-phases` too?

**Comment by JonathanStarup** [comment]
# Phase Benchmarks
## This PR:
Best runtime per phase (low to high):
  Deriver                          1,1 ms (00,0%)
  Reader                           1,7 ms (00,0%)
  Instances                        3,7 ms (00,1%)
  Terminator                       6,6 ms (00,1%)
  Safety                           8,7 ms (00,2%)
  Tailrec                          9,7 ms (00,2%)
  VarNumbering                    17,6 ms (00,3%)
  Lowering                        18,6 ms (00,4%)
  Simplifier                      21,6 ms (00,4%)
  Eraser                          21,7 ms (00,4%)
  Optimizer                       22,4 ms (00,4%)
  Stratifier                      24,4 ms (00,5%)
  LambdaLift                      25,1 ms (00,5%)
  Kinder                          30,2 ms (00,6%)
  Weeder                          32,5 ms (00,6%)
  PatternExhaustiveness           46,7 ms (00,9%)
  ClosureConv                     56,4 ms (01,1%)
  TreeShaker                      78,4 ms (01,5%)
  Namer                           91,3 ms (01,8%)
  Redundancy                      98,1 ms (01,9%)
  Finalize                       125,1 ms (02,5%)
  Resolver                       269,9 ms (05,3%)
  Parser                         272,2 ms (05,3%)
  Monomorph                      742,3 ms (14,5%)
  JvmBackend                     1492,5 ms (29,2%)
  Typer                          1586,2 ms (31,1%)

Finished 14 iterations on 58.705 lines of code in 102 seconds.

## Master:
Best runtime per phase (low to high):
  Deriver                          1,0 ms (00,0%)
  Reader                           1,8 ms (00,0%)
  Instances                        3,3 ms (00,1%)
  Terminator                       6,2 ms (00,1%)
  Safety                           8,4 ms (00,2%)
  Tailrec                          9,6 ms (00,2%)
  VarNumbering                    16,5 ms (00,3%)
  Lowering                        20,6 ms (00,4%)
  Simplifier                      20,9 ms (00,4%)
  Stratifier                      21,9 ms (00,4%)
  Optimizer                       22,2 ms (00,4%)
  Eraser                          22,3 ms (00,4%)
  LambdaLift                      25,7 ms (00,5%)
  Weeder                          26,2 ms (00,5%)
  Kinder                          30,3 ms (00,6%)
  PatternExhaustiveness           46,6 ms (00,9%)
  ClosureConv                     54,8 ms (01,1%)
  TreeShaker                      72,1 ms (01,4%)
  Namer                           85,6 ms (01,7%)
  Redundancy                      93,5 ms (01,8%)
  Finalize                       123,5 ms (02,4%)
  Resolver                       239,7 ms (04,7%)
  Parser                         354,8 ms (07,0%)
  Monomorph                      713,0 ms (14,1%)
  JvmBackend                     1450,0 ms (28,6%)
  Typer                          1592,1 ms (31,4%)

Finished 14 iterations on 58.705 lines of code in 97 seconds.

**Comment by magnus-madsen** [comment]
Seems within the margin of error.

**Comment by magnus-madsen** [comment]
Ill check this out local and see how it interacts with `api.flix.dev`. One thing that is missing:

In `Weeder` we should:

- [ ] Strip empty *beginning* lines.
- [ ] Strip empty *ending* lines.
- [ ] Trim all lines (i.e. remove whitespace before and after the line).

See `visitDoc`

**Comment by magnus-madsen** [comment]
When I compile `JonathanStarup:document-parsing` locally and run `--doc` I don't see any documentation (`"doc"`) at all. I guess something is broken?

**Comment by JonathanStarup** [comment]
> When I compile `JonathanStarup:document-parsing` locally and run `--doc` I don't see any documentation (`"doc"`) at all. I guess something is broken?

It doesn't parse correctly inside namespace, my mistake

**Comment by JonathanStarup** [comment]
Parsing is painful... The error was that my copied files had `\n` instead of `\r\n` and that `\r\n` is parsed as two new lines so all doc comments where split into their lines where only the last is considered...

File: main/src/ca/uwaterloo/flix/language/phase/Weeder.scala [hunk 1]
```diff
@@ -2238,7 +2238,12 @@ object Weeder extends Phase[ParsedAst.Program, WeededAst.Program] {
   /**
     * Weeds the given documentation.
     */
-  private def visitDoc(doc0: ParsedAst.Doc): Ast.Doc = Ast.Doc(doc0.lines.toList, mkSL(doc0.sp1, doc0.sp2))
+  private def visitDoc(doc0: ParsedAst.Doc): Ast.Doc = {
+    val trimmedLines = doc0.lines.map(_.trim)
+    val trimmedBeginning = trimmedLines.dropWhile(_ == "")
+    val trimmedEnd = trimmedBeginning.reverse.dropWhile(_ == "").reverse
```
Comment (magnus-madsen) [inline main/src/ca/uwaterloo/flix/language/phase/Weeder.scala:2244]: takeWhile?

**Comment by JonathanStarup** [comment]
> In the file `target/api/api.json`.

I can't run `--doc` for some reason. `java.lang.RuntimeException: Unable to write to path '.\target\api\api.json'.` even with admin

**Comment by magnus-madsen** [comment]
> > In the file `target/api/api.json`.
> 
> I can't run `--doc` for some reason. `java.lang.RuntimeException: Unable to write to path '.\target\api\api.json'.` even with admin

Maybe you need to create the folder `target/api`.

**Comment by JonathanStarup** [comment]
> 

Yeah that works, should I add a ticket?

File: main/src/ca/uwaterloo/flix/language/phase/Weeder.scala (hunk 1 above)
Comment (JonathanStarup) [reply main/src/ca/uwaterloo/flix/language/phase/Weeder.scala:2244]: That would stop at the first empty line. I couldn't find a nice way other than this and my brief internet search found nothing better

**Comment by magnus-madsen** [comment]
I think it would be a one-line fix in `Documentor`.

**Comment by JonathanStarup** [comment]
Okay, RFR. when I run `--doc` theres is only 80/1400 ish empty doc comments

**Comment by JonathanStarup** [comment]
<details>
<summary>New Benchmarks</summary>
--THIS PR--
Best runtime per phase (low to high):
  Deriver                          1,1 ms (00,0%)
  Reader                           1,8 ms (00,0%)
  Instances                        3,1 ms (00,1%)
  Terminator                       9,4 ms (00,2%)
  Tailrec                         10,0 ms (00,2%)
  Safety                          10,8 ms (00,2%)
  VarNumbering                    17,9 ms (00,3%)
  Lowering                        21,7 ms (00,4%)
  Optimizer                       21,9 ms (00,4%)
  Eraser                          22,0 ms (00,4%)
  Simplifier                      22,0 ms (00,4%)
  PatternExhaustiveness           22,7 ms (00,4%)
  Stratifier                      25,8 ms (00,5%)
  LambdaLift                      26,6 ms (00,5%)
  Weeder                          34,7 ms (00,7%)
  Kinder                          36,1 ms (00,7%)
  ClosureConv                     53,8 ms (01,0%)
  TreeShaker                      73,8 ms (01,4%)
  Namer                           82,7 ms (01,6%)
  Redundancy                      90,9 ms (01,8%)
  Finalize                       121,5 ms (02,4%)
  Resolver                       265,9 ms (05,2%)
  Parser                         295,0 ms (05,7%)
  Monomorph                      709,2 ms (13,8%)
  JvmBackend                     1555,6 ms (30,3%)
  Typer                          1606,1 ms (31,2%)

Finished 14 iterations on 58.384 lines of code in 101 seconds.
--MASTER--
</details>

**Comment by magnus-madsen** [comment]
> Okay, RFR. when I run `--doc` theres is only 80/1400 ish empty doc comments

Nice. I will try it out at home later with `api.flix.dev`.

File: main/src/ca/uwaterloo/flix/language/phase/Weeder.scala (hunk 1 above)
Comment (magnus-madsen) [reply main/src/ca/uwaterloo/flix/language/phase/Weeder.scala:2244]: Ok, I understand.

**Comment by magnus-madsen** [comment]
It seems to work for classes and enums (👍 ) but for signatures (defs inside classes) I don't get any docs.

**Comment by magnus-madsen** [comment]
Maybe that is those 40?

**Comment by magnus-madsen** [comment]
If it is difficult to fix, we can merge this in as-is.

**Comment by JonathanStarup** [comment]
> If it is difficult to fix, we can merge this in as-is.

that was an oversight, fixed now

**Comment by magnus-madsen** [comment]
Ill test when I am back home again.

**Comment by magnus-madsen** [comment]
Nice! Thanks

**Pull request merged**